### PR TITLE
builtins: Speed up set unions

### LINKED ISSUE #4979: union() is slower than pure Rego, can be up to O(n^3)
`union()` is very slow in Rego because of how it is implemented currently.

Consider the implementation of the `union` builtin:

```go
// builtinSetUnion returns the union of the given input sets
func builtinSetUnion(a ast.Value) (ast.Value, error) {

	inputSet, err := builtins.SetOperand(a, 1)
	if err != nil {
		return nil, err
	}

	result := ast.NewSet()

	err = inputSet.Iter(func(x *ast.Term) error {
		n, err := builtins.SetOperand(x.Value, 1)
		if err != nil {
			return err
		}
		result = result.Union(n)
		return nil
	})
	return result, err
}
```

And also `ast.set.Union()`:

```go
// Union returns the set containing all elements of s and other.
func (s *set) Union(other Set) Set {
	r := NewSet()
	s.Foreach(func(x *Term) {
		r.Add(x)
	})
	other.Foreach(func(x *Term) {
		r.Add(x)
	})
	return r
}
```

If we have $n$ many sets to union, of size $k$ each, then that's going to result in $n$ many calls to `ast.set.Union()`, except that each call to `ast.set.Union()` is going to end up creating a whole new set and copying all of the existing elements over to it.

This means that each call to `ast.set.Union()` takes increasingly more time, specifically $O(k), O(2k), O(3k), \dots$.
This means the running time is O(k * nth triangle number) = $O\left( \frac{(n+1)!}{2!(n-2)!} \cdot k\right)  = O\left(\frac{1}{2} k n (n^2-1)\right) \approx O(kn^3)$.

# Empirical Benchmarks

With the `union()` builtin:

```plain
$ cat slowunion/slowunion.rego
package slowunion

SIZE := 250
NSETS := 250

nums := numbers.range(0, NSETS)
sizes := numbers.range(0, SIZE)

sets[n] = x {
	nums[n]
	x := {sprintf("%d,%d", [n, i]) | sizes[i]}
}

combined := union({s | s := sets[_]})
$ time opa eval --profile --format pretty --bundle ./slowunion/ --import "data.slowunion" 'count(data.slowunion.combined)'
63001
+------------------------------+------------+
|            METRIC            |   VALUE    |
+------------------------------+------------+
| timer_rego_load_bundles_ns   | 583584     |
| timer_rego_module_compile_ns | 730417     |
| timer_rego_module_parse_ns   | 162458     |
| timer_rego_query_compile_ns  | 52125      |
| timer_rego_query_eval_ns     | 2673904333 |
| timer_rego_query_parse_ns    | 29625      |
+------------------------------+------------+
+--------------+----------+----------+-----------------------------+
|     TIME     | NUM EVAL | NUM REDO |          LOCATION           |
+--------------+----------+----------+-----------------------------+
| 2.543266402s | 4        | 254      | slowunion/slowunion.rego:14 |
| 115.799818ms | 63503    | 126253   | slowunion/slowunion.rego:11 |
| 3.801167ms   | 2        | 2        | data.slowunion.combined     |
| 152.332µs    | 3        | 3        | slowunion/slowunion.rego:6  |
| 119.959µs    | 1        | 251      | slowunion/slowunion.rego:10 |
| 95.999µs     | 3        | 3        | slowunion/slowunion.rego:7  |
| 4.624µs      | 1        | 1        | slowunion/slowunion.rego:4  |
| 2.125µs      | 1        | 1        | slowunion/slowunion.rego:3  |
+--------------+----------+----------+-----------------------------+

real	0m2.699s
user	0m3.573s
sys	0m0.137s
```

With pure Rego instead:

```plain
$ cat slowunion/slowunion.rego
package slowunion

SIZE := 250
NSETS := 250

nums := numbers.range(0, NSETS)
sizes := numbers.range(0, SIZE)

sets[n] = x {
	nums[n]
	x := {sprintf("%d,%d", [n, i]) | sizes[i]}
}

#combined := union({s | s := sets[_]})
combined := {t | s := sets[_]; s[t]}
$ time opa eval --profile --format pretty --bundle ./slowunion/ --import "data.slowunion" 'count(data.slowunion.combined)'
63001
+------------------------------+-----------+
|            METRIC            |   VALUE   |
+------------------------------+-----------+
| timer_rego_load_bundles_ns   | 538083    |
| timer_rego_module_compile_ns | 725375    |
| timer_rego_module_parse_ns   | 176458    |
| timer_rego_query_compile_ns  | 49208     |
| timer_rego_query_eval_ns     | 327751125 |
| timer_rego_query_parse_ns    | 31375     |
+------------------------------+-----------+
+--------------+----------+----------+-----------------------------+
|     TIME     | NUM EVAL | NUM REDO |          LOCATION           |
+--------------+----------+----------+-----------------------------+
| 195.567117ms | 254      | 63254    | slowunion/slowunion.rego:15 |
| 114.836799ms | 63503    | 126253   | slowunion/slowunion.rego:11 |
| 2.148999ms   | 2        | 2        | data.slowunion.combined     |
| 160.375µs    | 3        | 3        | slowunion/slowunion.rego:6  |
| 117.471µs    | 1        | 251      | slowunion/slowunion.rego:10 |
| 110.126µs    | 3        | 3        | slowunion/slowunion.rego:7  |
| 10.416µs     | 1        | 1        | slowunion/slowunion.rego:4  |
| 2.084µs      | 1        | 1        | slowunion/slowunion.rego:3  |
+--------------+----------+----------+-----------------------------+

real	0m0.354s
user	0m0.426s
sys	0m0.029s
```

System info:

```plain
$ uname -a
Darwin styra-macbook.local 21.5.0 Darwin Kernel Version 21.5.0: Tue Apr 26 21:08:37 PDT 2022; root:xnu-8020.121.3~4/RELEASE_ARM64_T6000 arm64
$ system_profiler SPSoftwareDataType SPHardwareDataType

Software:

    System Software Overview:

      System Version: macOS 12.4 (21F79)
      Kernel Version: Darwin 21.5.0
      Boot Volume: Macintosh HD
      Boot Mode: Normal
      Computer Name: styra-macbook
      User Name: Charles Daniels (cad)
      Secure Virtual Memory: Enabled
      System Integrity Protection: Enabled
      Time since boot: 5 days 15:17

Hardware:

    Hardware Overview:

      Model Name: MacBook Pro
      Model Identifier: MacBookPro18,3
      Chip: Apple M1 Pro
      Total Number of Cores: 8 (6 performance and 2 efficiency)
      Memory: 32 GB
      System Firmware Version: 7459.121.3
      OS Loader Version: 7459.121.3
      Serial Number (system): REDACTED
      Hardware UUID: REDACTED
      Provisioning UDID: REDACTED
      Activation Lock Status: Disabled

$ opa version
Version: 0.43.0
Build Commit: d75bbdd
Build Timestamp: 2022-07-29T21:15:42Z
Build Hostname: Mac-1659129125033.local
Go Version: go1.18.4
Platform: darwin/arm64
WebAssembly: unavailable
```

### PR DESCRIPTION
**Pull request by philipaconrad**
This PR fixes a performance regression for the set `union` builtin, discovered in issue #4979.

The original logic for the builtin did pairwise `Set.Union` calls between the input sets, resulting in a lot of wasted temporary Sets that were almost immediately discarded, and a lot of duplicated work. The updated builtin inlines the logic from `Set.Union`, so that only one pass is made across the incoming sets' members.

In the included benchmarks, this provides a small, but significant improvement over using a pure-Rego comprehension to do the same job.

Fixes #4979 

The 3x relevant benchmarks are shown below:
 - `union` on `main` branch
 - Pure Rego implementation (set comprehension)
 - `union` on this PR

## Benchmarks

Benchmark of `union` on `main` branch:
```
Running tool: /usr/local/go/bin/go test -benchmem -run=^$ -bench ^BenchmarkSetUnion$ github.com/open-policy-agent/opa/topdown

goos: linux
goarch: amd64
pkg: github.com/open-policy-agent/opa/topdown
cpu: Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz
BenchmarkSetUnion/10x10-8         	    2229	    544644 ns/op	  232683 B/op	    3759 allocs/op
BenchmarkSetUnion/10x100-8        	     265	   4402486 ns/op	 1754574 B/op	   24439 allocs/op
BenchmarkSetUnion/10x250-8        	     100	  11489154 ns/op	 4423815 B/op	   60298 allocs/op
BenchmarkSetUnion/100x10-8        	      62	  19831002 ns/op	 7055828 B/op	   32975 allocs/op
BenchmarkSetUnion/100x100-8       	       5	 216948499 ns/op	65802438 B/op	  228261 allocs/op
BenchmarkSetUnion/100x250-8       	       2	 639396404 ns/op	159630756 B/op	  570903 allocs/op
BenchmarkSetUnion/250x10-8        	       9	 117371750 ns/op	38499626 B/op	   86745 allocs/op
BenchmarkSetUnion/250x100-8       	       1	1469090944 ns/op	358555656 B/op	  622447 allocs/op
BenchmarkSetUnion/250x250-8       	       1	4215347082 ns/op	963369008 B/op	 1594872 allocs/op
PASS
ok  	github.com/open-policy-agent/opa/topdown	16.348s
```

Benchmark of Pure Rego implementation:
```
Running tool: /usr/local/go/bin/go test -benchmem -run=^$ -bench ^BenchmarkSetUnionSlow$ github.com/open-policy-agent/opa/topdown

goos: linux
goarch: amd64
pkg: github.com/open-policy-agent/opa/topdown
cpu: Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz
BenchmarkSetUnionSlow/10x10-8         	    2569	    461347 ns/op	  194384 B/op	    4015 allocs/op
BenchmarkSetUnionSlow/10x100-8        	     343	   4485940 ns/op	 1403187 B/op	   27487 allocs/op
BenchmarkSetUnionSlow/10x250-8        	     130	   8754293 ns/op	 3479220 B/op	   68079 allocs/op
BenchmarkSetUnionSlow/100x10-8        	     243	   4330704 ns/op	 1632044 B/op	   33589 allocs/op
BenchmarkSetUnionSlow/100x100-8       	      34	  33604653 ns/op	12636059 B/op	  244572 allocs/op
BenchmarkSetUnionSlow/100x250-8       	      13	  89953012 ns/op	31412245 B/op	  610317 allocs/op
BenchmarkSetUnionSlow/250x10-8        	     100	  10220165 ns/op	 4017694 B/op	   82992 allocs/op
BenchmarkSetUnionSlow/250x100-8       	      12	 101469788 ns/op	31077606 B/op	  606872 allocs/op
BenchmarkSetUnionSlow/250x250-8       	       3	 367955598 ns/op	80279797 B/op	 1514052 allocs/op
PASS
ok  	github.com/open-policy-agent/opa/topdown	13.738s
```

Benchmark of `union` on this PR:
```
PR:

Running tool: /usr/local/go/bin/go test -benchmem -run=^$ -bench ^BenchmarkSetUnion$ github.com/open-policy-agent/opa/topdown

goos: linux
goarch: amd64
pkg: github.com/open-policy-agent/opa/topdown
cpu: Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz
BenchmarkSetUnion/10x10-8         	    2788	    385342 ns/op	  174161 B/op	    3593 allocs/op
BenchmarkSetUnion/10x100-8        	     433	   2946532 ns/op	 1224484 B/op	   24095 allocs/op
BenchmarkSetUnion/10x250-8        	     123	   9764934 ns/op	 3036327 B/op	   59735 allocs/op
BenchmarkSetUnion/100x10-8        	     313	   3304217 ns/op	 1442761 B/op	   29577 allocs/op
BenchmarkSetUnion/100x100-8       	      45	  27416137 ns/op	10991659 B/op	  213286 allocs/op
BenchmarkSetUnion/100x250-8       	      15	  73213722 ns/op	27342127 B/op	  533573 allocs/op
BenchmarkSetUnion/250x10-8        	     138	   8663335 ns/op	 3553444 B/op	   72984 allocs/op
BenchmarkSetUnion/250x100-8       	      16	  71706943 ns/op	26999915 B/op	  529107 allocs/op
BenchmarkSetUnion/250x250-8       	       6	 199493126 ns/op	70179568 B/op	 1323346 allocs/op
PASS
ok  	github.com/open-policy-agent/opa/topdown	15.198s
```

## Summary of Results

The benchmarks show that the PR is roughly 15-30% faster and has ~14% fewer allocations than the pure Rego solution. The resolution of these charts are blown out by the 10x10 and 250x250 cases, but still paint a strong picture of the performance changes nonetheless:

2022-08-08_set-union-optimization

### CODE REVIEW

File: topdown/sets.go [hunk 1]
```diff
@@ -58,22 +58,28 @@ func builtinSetIntersection(a ast.Value) (ast.Value, error) {
...
 	err = inputSet.Iter(func(x *ast.Term) error {
-		n, err := builtins.SetOperand(x.Value, 1)
+		item, err := builtins.SetOperand(x.Value, 1)
 		if err != nil {
 			return err
 		}
-		result = result.Union(n)
+		item.Foreach(func(y *ast.Term) {
+			result.Add(y)
+		})
```
Comment (srenatus) [inline topdown/sets.go:79]: [golf] Isn't this the same as

```suggestion
		item.Foreach(result.Add)
```

One of those few lucky moments when the signatures _just fit_ 😅

Comment (philipaconrad) [reply topdown/sets.go:79]: Oooooh! I think you might be right! :smiley: I'll test that shortly.

Comment (philipaconrad) [reply topdown/sets.go]: Comitted that exact suggestion.

**Comment by philipaconrad** [comment]
I'm going to add some correctness/regression tests, at the suggestion of @charlesdaniels.

The tests will compare the behavior of `union(...)` against pairwise `Set.Union` operations for various lists of input sets. These should be cheap to run, and will help us prevent a silly regression if our existing test suite doesn't already check for this stuff.

**Comment by philipaconrad** [comment]
I've split the `union` tests/benchmarks off into their own files, since I'm planning to soon be doing the same process for the `intersection` builtin, and `topdown_bench_test` was getting a bit crowded.

**Comment by philipaconrad** [comment]
@tsandall @srenatus This PR should be ready for review! :smile:

**Review by tsandall — APPROVED** [review]

**Comment by tsandall** [comment]
LGTM. In this case, inlining the union logic is cheap enough that the speedup is worth it.

**Comment by philipaconrad** [comment]
@tsandall Thanks for looking the PR over! I'll merge this in soon.

**Pull request merged**